keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+
+Each generated `Task` is expected to target a specific `Node` in the cluster.
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Task
+metadata:
+  name: task-sample-control-plane-1
+  labels:
+    app: a
+spec:
+  nodeName: control-plane-1
+  alphaCertsRenew: {}
+```
+
+or
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Task
+metadata:
+  name: task-sample-control-plane-2
+  labels:
+    app: a
+spec:
+  nodeName: control-plane-2
+  alphaCertsRenew: {}
+```
+
+The `TaskGroup controller` is going to generate `Task` according to a policy that by default is sequential order - only after the current `Task` is completed successfully a new one will be generated.
+
+In future releases of the kubeadm operator also additional policy could be added for supporting e.g. parallel execution of `Task` or disruption budget (max N `Task` in executed at the same time).
+
+In any case, the `TaskGroup controller` is going to ensure a predictable execution order based on the alphabetical order of `Node`s in the scope of a `TaskGroup`.
+
+#### TaskController
+
+The `Task controller` is the last component of the kubeadm operator. 
+
+The main characteristic of this controller is the fact that this controller is deployed on all the nodes in the cluster using a `DaemonSet`, and that each instance is responsible for managing `Task` targeting a specific node.
+
+`Task` reconciliation triggers the execution of kubeadm commands on such node; in order to do so, `Task controller` require to be executed in privileged mode.
+
+#### Observability and Operability
+The kubeadm operator is designed to take responsibility for actions that as of today are in charge of users, and we want to ensure that the users can feel comfortable in handling over such responsibility.
+
+This goal influences the kubeadm operator design as described in the following paragraphs
+
+##### Execution order
+
+The kubeadm operator is going to execute `Action`'s `Task`s in a predictable and consistent order:
+
+- `TaskGroup` will be generated in the order encoded in each action and always executed in sequential order.
+- If a `TaskGroup` uses `sequential` policy, `Task` will be executed according to the alphabetical order of Nodes.
+- Additional `TaskGroup` policies introduced in future, are expected to introduce similar guarantees. 
+
+Additionally, it will be possible to execute actions in `dry-run` mode, thus allowing users to have a preview of the resulting task order.
+
+##### Execution modes
+
+The kubeadm operator is going to execute`Task`s acconding to the order described above, and by default a new `Task` will be automatically started as soon as the previous task completes.
+
+At any time during the execution, the user will be allowed to pause the `Action` execution using the new `kubeadm operator pause` command.
```

Comment:
```suggestion
At any time during the execution, the user will be allowed to pause the next `Action` execution using the new `kubeadm operator pause` command.
```

Reply:
Actually the idea is to not pause the next action, but the next task within the same action.
I rephrased to make this more clear